### PR TITLE
Fix #396, #395: Use typed exceptions for minio client

### LINKED ISSUE #396: [Suggestion] Fixes for MinioException.
The MinioException has two public properties, `message` and `response`.

I'd suggest to use standard `Exception.Message` property. Client code may log only ex.Message. Having an empty message mages troubleshooting much harder. As it is for now, the `MinioException.Message` is null for some exceptions, BucketNotFoundException as example.

The `response` property should be named in PascalCase co conform c# naming guidelines. In addition, I believe it should be set for all exceptions where possible. Especially for the InternalClientException.
Currently there is no way to distinguish "server unavailable" error from "invalid object name" error. Both of them throw same InternalClientException with same properties.

### PR DESCRIPTION
**Pull request by ig-sinicyn**
* Rename MinioException properties: message => ServerMessage; response => ServerResponse.
* Split ParseError method into simpler and shorter methods
* Throw exceptions for invalid object name and connectivity errors.
* Add NegativeTest for newly added exceptions
* Pass IRestResponse to connectivity-related and unexpected-error exceptions. 
* Add SelectObjectContentException and UnexpectedMinioException exceptions. Code now use specific exceptions and do not throw base MinioException anymore.

### CODE REVIEW

**Comment by ig-sinicyn** [comment]
Up :)

File: Minio.Tests/NegativeTest.cs [hunk 1]
```diff
@@ -0,0 +1,72 @@
+﻿/*
+ * MinIO .NET Library for Amazon S3 Compatible Cloud Storage, (C) 2017 MinIO, Inc.
```
Comment (Praveenrajmani) [inline Minio.Tests/NegativeTest.cs:2]: We have to fix the YEAR here. In other places as well.

Comment (ig-sinicyn) [reply Minio.Tests/NegativeTest.cs:2]: Ok, will do

Comment (ig-sinicyn) [reply Minio.Tests/NegativeTest.cs:2]: Done:)

Comment (poornas) [reply Minio.Tests/NegativeTest.cs]: @ig-sinicyn, the year needs to be fixed only on the files touched by your PR. Could you revert the changes to rest of the files?

Comment (ig-sinicyn) [reply Minio.Tests/NegativeTest.cs]: Done:)

**Pull request merged**